### PR TITLE
Add new indexer TorrentParadise. resolves #4598

### LINKED ISSUE #4598: New tracker request - https://torrent-paradise.ml/
New decentralised tracker appear last days, is it possible to add it in Jackett ?

Info : https://torrentfreak.com/torrent-paradise-creates-decentralized-pirate-bay-with-ipfs-190120/

Tracker website : https://torrent-paradise.ml/

### PR DESCRIPTION
**Pull request by XYZJR**
https://github.com/Jackett/Jackett/issues/4598

- Why does adding the indexer immediately call PerformQuery to test the indexer, with an empty querystring? This indexer doesn't have a default results page, so I'm not sure if my solution (search for the word "test") is OK. When manually running a test on the indexer, it uses an empty querystring once more.
- This indexer has no categories. Which ones should I add if it has a bit of everything?

### CODE REVIEW

**Comment by XYZJR** [comment]
@cadatoiva How can I add you as a reviewer? Still figuring things out on Github.

**Comment by ilike2burnthing** [comment]
There's already an indexer with the ID [torrentparadise](https://github.com/Jackett/Jackett/blob/master/src/Jackett.Common/Definitions/torrentparadise.yml), so you'll need to rename this to something like `torrent-paradise` or `torrentparadise-ml`.

If a tracker doesn't have a recent page, and a keywordless search doesn't work, then we use the current year. For YAML indexers this is done with `.Today.Year`, but I'm not sure how that's done in C#.

If there are no categories then use `Other` (8000).

DL and UL Factor should both be 1 for public trackers.

If a tracker doesn't have details pages for results then use the sitelink.

If a tracker doesn't have upload dates then use 0mins/now.

You'll also need to add `Torrent Paradise` to the list of supported public trackers in the [README](https://github.com/Jackett/Jackett/blob/master/README.md).

**Comment by XYZJR** [comment]
@ilike2burnthing

I didn't check the YAML indexers to see if my indexer ID was unique. I expected an error to be thrown if there are duplicate IDs.

Can you tell me which team members are still actively working on C# code?

Should I add that `Other` category to each individual result as well, or just the indexer?

Any idea what `BaseIndexer.Encoding` is used for? I've copied it from another indexer but I haven't figured out where it's used yet.

**Comment by ilike2burnthing** [comment]
Just @ngosang at the moment, though occasionally we get PRs from others such as yourself.

All results should have a category, so they should be `Other` as well.

`Encoding` is the character set used by the site, which in this case is `UTF-8`.

**Comment by XYZJR** [comment]
@ngosang @ilike2burnthing Any news on this? I want to know if I'm doing this right before I move on to another indexer or bug fix.

**Comment by ilike2burnthing** [comment]
Just waiting on ngosang's review when they're free.

File: README.md
Comment (ngosang) [inline README.md:134]: Torrent Paradise (ML)

File: src/Jackett.Common/Indexers/TorrentParadise.cs
Comment (ngosang) [inline src/Jackett.Common/Indexers/TorrentParadise.cs:18]: Make this class private and move it to the bottom

File: src/Jackett.Common/Indexers/TorrentParadise.cs
Comment (ngosang) [inline src/Jackett.Common/Indexers/TorrentParadise.cs:47]: torrent-paradise-ml

File: src/Jackett.Common/Indexers/TorrentParadise.cs
Comment (ngosang) [inline src/Jackett.Common/Indexers/TorrentParadise.cs:16]: Rename the class to `TorrentParadiseMl`. The other TorrentParadise site is more popular.

File: src/Jackett.Common/Indexers/TorrentParadise.cs
Comment (ngosang) [inline src/Jackett.Common/Indexers/TorrentParadise.cs:48]: Torrent Paradise (ML)

File: src/Jackett.Common/Indexers/TorrentParadise.cs
Comment (ngosang) [inline src/Jackett.Common/Indexers/TorrentParadise.cs:88]: You don't need to set the year here.

File: src/Jackett.Common/Indexers/TorrentParadise.cs
Comment (ngosang) [inline src/Jackett.Common/Indexers/TorrentParadise.cs:99]: Remove the CreateSearchUrl method and use this code instead (not tested):
```
var searchTerm = query.GetQueryString();
if (string.IsNullOrWhiteSpace(searchTerm)) {
   if (query.IsTest)
         searchTerm = DateTime.Now.Year.ToString();
   else 
         return new List<ReleaseInfo>();
}
 var qc = new List<KeyValuePair<string, string>>
{
    {"q", searchTerm}
};
var searchUrl = ApiUrl + "search?" + qc.GetQueryString();
```

File: src/Jackett.Common/Indexers/TorrentParadise.cs
Comment (ngosang) [inline src/Jackett.Common/Indexers/TorrentParadise.cs:148]: Remove the method `CreateMagnetUri`
If you set InfoHash = result.Id the magnet will be auto-generated.

File: src/Jackett.Common/Indexers/TorrentParadise.cs
Comment (ngosang) [inline src/Jackett.Common/Indexers/TorrentParadise.cs:77]: Encoding = Encoding.UTF8;

Comment (ngosang) [inline src/Jackett.Common/Indexers/TorrentParadise.cs:77]: Encoding = Encoding.UTF8;

**Review by ngosang — CHANGES_REQUESTED** [review]
Thank you! I let you some changes. It can be merged after they are resolved.

File: src/Jackett.Common/Indexers/TorrentParadise.cs
Comment (XYZJR) [reply src/Jackett.Common/Indexers/TorrentParadise.cs:99]: I'd like to keep it as a method since it does its own specific task and it also keeps `PerformQuery` less cluttered.
I've changed it a bit, as you suggested. I still update the `query.SearchTerm`, which keeps everything in sync (for when you might use `TorznabQuery query` later, in `PerformQuery`).

Comment (ngosang) [reply src/Jackett.Common/Indexers/TorrentParadise.cs]: @XYZJR There are few things I didn't tell you. You can keep the method, it's fine.
1) You shouldn't modify the query object because it's shared by other indexers. In Jackett you can make a search in several indexers at the same time and all of them receive the same query object. This is a bug in Jackett. I tried to fix it in this PR but I think it should be a better fix based on immutable objects. Maybe you can take a look => #8570
2) Replace `List<KeyValuePair<string, string>>` with `NameValueCollection`.  My bad. We are using this object to concatenate query parameters because `qc.GetQueryString()` escapes the keys and values to form a valid URL

Comment (XYZJR) [reply src/Jackett.Common/Indexers/TorrentParadise.cs]: Alright, changing it now.

PS: Is there an easier way to get in contact with you, are you on the Jackett Discord for example?

Comment (ngosang) [reply src/Jackett.Common/Indexers/TorrentParadise.cs]: email ngosang at hotmail.es but the time to response is similar to github.
I will try to review your PRs faster but I'm very busy, sorry.

**Pull request merged**
